### PR TITLE
Feature: add cache and common unit tests

### PR DESCRIPTION
**Pull request by MurashovVen**
- Adding cache unit tests.
- Adding common unit tests.
- Minor code fixes: 
adding error checking;
replacing `defer` calling;

### CODE REVIEW

File: code/go/0chain.net/core/common/rate_limter_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+			viper.Set("network.n2n_handlers.rate_limit", tt.n2nRL)
+			ConfigRateLimits()
+
+			w := httptest.NewRecorder()
+			r := httptest.NewRequest(http.MethodGet, "/", nil)
+			handler := N2NRateLimit(tt.args.handler)
+			handler(w, r)
+			w.Body = nil
+
+			if !reflect.DeepEqual(w, tt.want) {
```
Comment (Sriep) [inline code/go/0chain.net/core/common/rate_limter_test.go:127]: `reflect.DeepEqual` seems to return different values in different build modes. I think the `w.HeaderMap` is empty and some bug in `reflect` is causing this to be missed in release mode. Anyway recheck `w.HeaderMap` and `tt.want.HeaderMap` are as expected.

**Review by Sriep — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/core/common/mdw_recover_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+			}(),
+			UseRecoveredHandler: true,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			UseRecoverHandler = tt.UseRecoveredHandler
```
[the remark this answers is not in the document]
Comment (MurashovVen) [reply code/go/0chain.net/core/common/mdw_recover_test.go:100]: This is not a build constant, it is a global dynamic variable.
We also have `UseRecoverHandler=false` test on line 21. We can remove parallel running for these tests or make the `UseRecoverHandler` constant, or protect it against concurrent access. If we just move assignment out of the test, tests will fall.

File: code/go/0chain.net/core/common/rate_limter_test.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+			want:   w,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			viper.Set("network.user_handlers.rate_limit", tt.userRL)
+			ConfigRateLimits()
```
[the remark this answers is not in the document]
Comment (MurashovVen) [reply code/go/0chain.net/core/common/rate_limter_test.go:58]: This is where tests run in parallel with different configurations. If we move the configuration setup before the test, the tests will fail.
We can remove parallel running from these tests, or make config setup protected against concurrent access.

File: code/go/0chain.net/core/common/rate_limter_test.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+			want:  w,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			viper.Set("network.n2n_handlers.rate_limit", tt.n2nRL)
+			ConfigRateLimits()
```
[the remark this answers is not in the document]
Comment (MurashovVen) [reply code/go/0chain.net/core/common/rate_limter_test.go:119]: This is where tests run in parallel with different configurations. If we move the configuration setup before the test, the tests will fail.
We can remove parallel running from these tests, or make config setup protected against concurrent access.

File: code/go/0chain.net/core/common/rate_limter_test.go (hunk 1 above)
Comment (MurashovVen) [reply code/go/0chain.net/core/common/rate_limter_test.go]: Added checks for `w.HeaderMap`.

File: code/go/0chain.net/core/common/mdw_recover_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sriep) [reply code/go/0chain.net/core/common/mdw_recover_test.go:100]: I mean a logical constant; It is set once in [file startup](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/core/common/mdw_recover.go#L13) and never changed.

File: code/go/0chain.net/core/common/rate_limter_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Sriep) [reply code/go/0chain.net/core/common/rate_limter_test.go:58]: ConfigRateLimits is only called once, in [miner startup](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/miner/miner/main.go#L206) or in [shrader startup](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/sharder/sharder/main.go#L260).

To simulate I think the best place is in init.
```go
func init() {
   //... anything configRateLinits wants...
  
  ConfigRateLimits()
}
```
Confounding race errors from spf13/viper is making it difficult to see what's going on, this should be fixed in Pr https://github.com/0chain/0chain/pull/246.

The tests should be kept in parallel, as handlers run in parallel.

File: code/go/0chain.net/core/common/rate_limter_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sriep) [reply code/go/0chain.net/core/common/rate_limter_test.go:119]: As above I think ConfigRateLimitis belongs in `init`?

File: code/go/0chain.net/core/common/rate_limter_test.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+			w := httptest.NewRecorder()
+			r := httptest.NewRequest(http.MethodGet, "/", nil)
+			handler := N2NRateLimit(tt.args.handler)
+			handler(w, r)
+			w.Body = nil
+
+			if !reflect.DeepEqual(w.HeaderMap, tt.want.HeaderMap) && !reflect.DeepEqual(w, tt.want) {
```
Comment (Sriep) [inline code/go/0chain.net/core/common/rate_limter_test.go:127]: [httptest.ResponseRecorder.Headermap](https://golang.org/pkg/net/http/httptest/#ResponseRecorder) is Deperciated
```go
    // HeaderMap contains the headers explicitly set by the Handler.
    // It is an internal detail.
    //
    // Deprecated: HeaderMap exists for historical compatibility
    // and should not be used. To access the headers returned by a handler,
    // use the Response.Header map as returned by the Result method.
    HeaderMap http.Header
```

File: code/go/0chain.net/core/common/mdw_recover_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MurashovVen) [reply code/go/0chain.net/core/common/mdw_recover_test.go]: I've removed test with `UseRecoverHandler=false` and added parallel running.

File: code/go/0chain.net/core/common/rate_limter_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MurashovVen) [reply code/go/0chain.net/core/common/rate_limter_test.go]: I've removed test with `0` rate limit, moved rate limit initializing in `init` function and added parallel running.

File: code/go/0chain.net/core/common/rate_limter_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MurashovVen) [reply code/go/0chain.net/core/common/rate_limter_test.go]: I've removed test with `0` rate limit, moved rate limit initializing in `init` function and added parallel running.

File: code/go/0chain.net/core/common/rate_limter_test.go (hunk 5 above)
Comment (MurashovVen) [reply code/go/0chain.net/core/common/rate_limter_test.go]: Replaced on `Header()` result checking.

**Review by Sriep — APPROVED** [review]

**Pull request merged**